### PR TITLE
Add utility functions for dealing with PDF.k units

### PR DESCRIPTION
**Pull request by EmperorArthur**
* Added unit tests for all of the units accepted by `fpdf.FPDF`
* Added a `get_scale_factor` utility function for obtaining information about a unit.
* Adjusted `fpdf.FPDF` to raise a ValueError when passed an incorrect unit.
* Added a `convert_unit` utility function to allow for unit conversion. Including complex data like a list of points.
* Adjusted `fpdf.FPDF` to accept a number as a unit.

That exception one was done so the utility functions can be kept lean, and because that's the normal exception thrown in cases like that.

Similarly, allowing a number as a unit simplifies things significantly.  For example, data from one instance of `fpdf.FPDF` can now be used to create another instance with less work.  Previously, the other instance would have to be created, then "k" changed manually.

The primary benefit is unit conversion.  Which can be critical when writing generic code designed to interact with the system.  Another example is trying to create Templates, but live in the US where everything is in inches.

### Edit:

A previous version contained a `to_unit` function.  On further review, I found that this was merely a special case of `convert_unit`, and was not needed.

### CODE REVIEW

**Comment by EmperorArthur** [comment]
Those tests are failing on code that I did not commit!

Also, if you don't like the ValueError thing, it can be caught and re-thrown as the correct type.

**Comment by Lucas-C** [comment]
Looks awesome, thank you for your contribution!

I'll review it next week.

**Comment by EmperorArthur** [comment]
Changes have been made.  Ready for review.

**Comment by EmperorArthur** [comment]
I am glad to contribute and hope everything meets your standards and project style.  If it does not, I will be happy to change things.

As someone who has spent many hours "enjoying" the original FPDF, I really appreciate this version and would love to make it even better.

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -195,7 +196,7 @@ def __init__(
         Args:
             orientation (str): possible values are "portrait" (can be abbreviated "P")
                 or "landscape" (can be abbreviated "L"). Default to "portrait".
-            unit (str): possible values are "pt", "mm", "cm" or "in".
+            unit (str, int, float): possible values are "pt", "mm", "cm" or "in".
```
Comment (Lucas-C) [inline fpdf/fpdf.py:199]: Could you please explain in docstring how a numeral value is interpreted / used?

File: fpdf/util.py [hunk 2]
```diff
@@ -41,6 +42,54 @@ def b(s):
     raise ValueError(f"Invalid input: {s}")
 
 
+def get_scale_factor(unit: Union[str, float, int]) -> float:
+    """
+    Get how many pts are in a unit. (k)
+
+    Args:
+        unit (str, float, int): Any of "pt", "mm", "cm", "in", or a number
```
Comment (Lucas-C) [inline fpdf/util.py:50]: Same remark: could you please explain in docstring how a numeral value is interpreted / used by this function?
Judging from the code, a numeral value would directly represent a number of points, right?

File: test/errors/test_FPDF_errors.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def test_incorrect_orientation():
 
 
 def test_units():
-    with pytest.raises(FPDFException) as e:
+    with pytest.raises(ValueError) as e:
```
Comment (Lucas-C) [inline test/errors/test_FPDF_errors.py:43]: 👍

File: test/layout/test_units.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+def test_unit_default(tmp_path):
+    """Test creating a PDF with no unit"""
+    pdf = FPDF()
+    pdf.set_font("Helvetica")
+    pdf.add_page()
+    pdf.cell(w=10, h=10, txt="default")
+    assert_pdf_equal(pdf, HERE / "unit_default.pdf", tmp_path)
+
+
+def test_unit_adjustment(tmp_path):
```
Comment (Lucas-C) [inline test/layout/test_units.py:20]: Just a detail: could you please have the function name & reference PDF file name match?

File: test/layout/test_units.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    """Test creating a PDF with no unit, then changing to another works"""
+    pdf = FPDF()
+    pdf.set_font("Helvetica")
+    pdf.add_page()
+    pdf.cell(w=10, h=10, txt="default")
+    pdf.k = 1
+    pdf.add_page()
+    pdf.cell(w=10, h=10, txt="pt")
+    assert_pdf_equal(pdf, HERE / "unit_changed.pdf", tmp_path)
```
Comment (Lucas-C) [inline test/layout/test_units.py:29]: Seems like the PDf reveals that only the margins are affected by the change on the 2nd page, right?

File: test/layout/test_units.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    assert_pdf_equal(pdf, HERE / "unit_{}.pdf".format(unit), tmp_path)
+
+
+def test_unit_int(tmp_path):
+    """Test creating a PDF with an int unit"""
+    pdf = FPDF(unit=72)
+    pdf.set_font("Helvetica")
+    pdf.add_page()
+    pdf.cell(w=10, h=10, txt="in")
+    assert_pdf_equal(pdf, HERE / "unit_in.pdf", tmp_path)
```
Comment (Lucas-C) [inline test/layout/test_units.py:48]: Shouldn't the filename be `unit_int`?

File: test/layout/test_units.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def test_unit_float(tmp_path):
+    """Test creating a PDF with a float unit"""
+    pdf = FPDF(unit=72.0)
+    pdf.set_font("Helvetica")
+    pdf.add_page()
+    pdf.cell(w=10, h=10, txt="in")
+    assert_pdf_equal(pdf, HERE / "unit_in.pdf", tmp_path)
```
Comment (Lucas-C) [inline test/layout/test_units.py:57]: Shouldn't the filename be `unit_float`?

File: test/utils/test_unit_conversion.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+from fpdf.util import convert_unit
+
+
+def test_convert_unit_numbers():
+    """Test converting units where both old and new are numbers"""
+    converted = convert_unit(1, 72, 72 / 25.4)
+    assert (converted - 25.4) < 0.0001  # Account for floating point errors
```
Comment (Lucas-C) [inline test/utils/test_unit_conversion.py:7]: Alternatively, you could use `math.isclose`:
https://stackoverflow.com/a/33024979/636849

**Review by Lucas-C — CHANGES_REQUESTED** [review]
Really good job overall!
The tests are very complete 😊

I fixed the `pylint` checks on the `master` branch, so just rebasing yours on it should get rid of those errors.

Feel free to also add your name to the `CONTRIBUTORS.md` file as part of this PR.

File: test/layout/test_units.py (hunk 6 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:48]: That's deliberate.  Added comments to explain what's going on.

It's not just testing creation, it's also making sure that one created with an `int` or `float` are exactly the same as one created with `unit="in"`.

If this were Python 2, then this test would fail, since k would end up causing integer division when configuring the margins.

File: test/layout/test_units.py (hunk 7 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:57]: That's deliberate.  Added comments to explain what's going on.

It's not just testing creation, it's also making sure that one created with an `int` or `float` are exactly the same as one created with `unit="in"`.

File: test/layout/test_units.py (hunk 4 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:20]: Done

File: test/layout/test_units.py (hunk 6 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:48]: I've gone ahead and moved that to two different tests.

File: test/layout/test_units.py (hunk 7 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:57]: I've gone ahead and moved that to two different tests.

File: test/layout/test_units.py (hunk 5 above)
Comment (EmperorArthur) [reply test/layout/test_units.py:29]: The cell size should have also changed, but it's not noticeable.  However, the margins should not have changed, so I have commented out the test for now.

I have an idea for a fix, but want to do one thing at a time.

File: fpdf/util.py (hunk 2 above)
Comment (EmperorArthur) [reply fpdf/util.py]: Done

File: fpdf/fpdf.py (hunk 1 above)
Comment (EmperorArthur) [reply fpdf/fpdf.py]: Done

File: CHANGELOG.md [hunk 9]
```diff
@@ -7,6 +7,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/),
 and [PEP 440](https://www.python.org/dev/peps/pep-0440/).
 
+## [2.4.3] - not released yet
+### Added
+- `get_scale_factor` utility function to obtain FPDF.k without having to create a document.
+- `convert_unit` utility function to convert a number, x,y point, or list of x,y points from one unit to another unit
+
+### Changed
+- `fpdf.FPDF` constructor now returns a `ValueError` when an invalid unit is provided.
```
[the remark this answers is not in the document]
Comment (EmperorArthur) [reply CHANGELOG.md]: Done

File: test/utils/test_unit_conversion.py (hunk 8 above)
Comment (EmperorArthur) [reply test/utils/test_unit_conversion.py]: Done

**Comment by EmperorArthur** [comment]
Thank you for the feedback.  All of the changes you have requested should now be in there.

File: fpdf/fpdf.py [hunk 10]
```diff
@@ -195,9 +196,10 @@ def __init__(
         Args:
             orientation (str): possible values are "portrait" (can be abbreviated "P")
                 or "landscape" (can be abbreviated "L"). Default to "portrait".
-            unit (str): possible values are "pt", "mm", "cm" or "in".
+            unit (str, int, float): possible values are "pt", "mm", "cm", "in", or a number.
                 A point equals 1/72 of an inch, that is to say about 0.35 mm (an inch being 2.54 cm).
                 This is a very common unit in typography; font sizes are expressed in this unit.
+                If given a number, then it will be treated as the number of points per unit.  (eg. 72 = 1 in)
```
Comment (Lucas-C) [inline fpdf/fpdf.py:202]: 👍

File: test/conftest.py [hunk 11]
```diff
@@ -61,6 +61,7 @@ def assert_pdf_equal(actual, expected, tmp_path, generate=False):
             if isinstance(expected, (bytes, bytearray)):
                 pdf_file.write(expected)
             else:
+                expected.set_creation_date(EPOCH)
```
Comment (Lucas-C) [inline test/conftest.py:64]: Why is this needed?

**Review by Lucas-C — APPROVED** [review]
Looks good to me!

Could you expand on the fix you had in minde for `test_unit_adjustment`?

I'd de prefer if this could included as part of this PR if you are OK,
or else remove the commented-out unit test 🙂

File: test/layout/test_units.py (hunk 5 above)
Comment (EmperorArthur) [reply test/layout/test_units.py]: Test & Fix will be in a separate patch.

File: test/conftest.py (hunk 11 above)
Comment (EmperorArthur) [reply test/conftest.py]: See my other pull request.  Without this, the tests fail due to a bad creation date.

**Comment by EmperorArthur** [comment]
I fully expect the CI to fail, as without PR #186 comparing two FPDF classes is broken.

**Comment by Lucas-C** [comment]
Could you rebase this PR now that I have merged #186 please ?

**Comment by EmperorArthur** [comment]
Done

**Review by alexp1917 — COMMENTED** [review]
looks good to me, ill let Lucas merge since there are a number of comments above already and im just weighing in now

**Comment by Lucas-C** [comment]
Merged!

Thank you very much @EmperorArthur 😉

And thanks for the review @alexp1917!
